### ISSUE TITLE
Strict mode SyntaxError

### ISSUE THREAD

**Issue by epoberezkin** [MatAtBread/nodent#22]
```
SyntaxError: In strict mode code, functions can only be declared at top level or immediately within another function.
```

It happens in node 0.10 and 0.12, but doesn't happen in 4 and 5.

**Comment by epoberezkin** [comment]
this: `function a() { 'use strict'; if (true) { function b() {} } b(); }` is a syntax error in 0.10-12 and generates runtime error in 4 and 5 (`b is undefined`)

this: `function a() { 'use strict'; if (true) { b(); function b() {} } }` is also a syntax error in 0.10-12 and works ok in 4 and 5

**Comment by matAtWork** [comment]
I'm not sure this is a nodent bug - nodent transpiles async/await, re-ordering code where it has to, however if your source is syntactically incorrect (as the above example is), it won't fix it. 

Are you posting this issue because nodent is generating code like this, or because it doesn't fix it?

**Comment by epoberezkin** [comment]
Yes, nodent generates code like this.

For example this code:

```
validate = async function(data, dataPath) {
  'use strict';
  var vErrors = null;
  var errors = 0;
  if ((data && typeof data === "object" && !Array.isArray(data))) {
    var errs__0 = errors;
    var valid1 = true;
    if (data.word === undefined) {
      valid1 = true;
    } else {
      var errs_1 = errors;
      try {
        var valid1 = await refVal1(data.word, (dataPath || '') + '.word');
      } catch (e) {
        if (!(e instanceof ValidationError)) throw e;
        if (vErrors === null) vErrors = e.errors;
        else vErrors = vErrors.concat(e.errors);
        errors = vErrors.length;
      }
      var valid1 = errors === errs_1;
    }
  }
  if (errors === 0) return true;
  else throw new ValidationError(vErrors);
};
```

transpiled to this:

```
validate = function (data, dataPath) {
    return new Promise((function ($return, $error) {
        var vErrors, errors, errs__0, valid1, errs_1;
        'use strict';
        vErrors = null;
        errors = 0;
        function $IfStatement_2() {
            if (errors === 0) 
                return $return(true);
             else 
                return $error(new ValidationError(vErrors));
            return $return();
        }
        
        if (data && typeof data === "object" && !Array.isArray(data)) {
            errs__0 = errors;
            valid1 = true;
            function $IfStatement_3() {
                return $IfStatement_2.call(this);
            }
            
            if (data.word === undefined) {
                valid1 = true;
                return $IfStatement_3.call(this);
            } else {
                var $Try_1_Catch = (function (e) {
                    if (!(e instanceof ValidationError)) 
                        return $error(e);
                    if (vErrors === null) 
                        vErrors = e.errors;
                     else 
                        vErrors = vErrors.concat(e.errors);
                    errors = vErrors.length;
                    $Try_1_Post.call(this)
                }).$asyncbind(this, $error);
                function $Try_1_Post() {
                    valid1 = errors === errs_1;
                    return $IfStatement_3.call(this);
                }
                
                errs_1 = errors;
                try {
                    return refVal1(data.word, (dataPath || '') + '.word').then((function ($await_4) {
                        valid1 = $await_4;
                        $Try_1_Post.call(this)
                    }).$asyncbind(this, $Try_1_Catch), $Try_1_Catch);
                } catch (e) {
                    $Try_1_Catch(e)
                }
            }
        }
        return $IfStatement_2.call(this);
    }).$asyncbind(this));
};
```

that has functions defined in blocks.

**Comment by matAtWork** [comment]
This is an interesting one. It appears the use of FunctionDeclaration within a block is strictly not part of the language, but all browsers allow it. In ES5 they are hoisted like 'var' to the top of the containing function scope. In ES5-Strict, they are 'syntax errors', as per the spec, partially (at least) to make way for the ES6 behaviour when they are hoisted to the containing block. They can be replaced with FunctionExpressions, but these are considerably slower in V8. 

This means nodent would have to know the execution environment to place them correctly (ES5, ES5-Strict or ES6), which would usually mean falling back to ES5-Strict (hoisted to top of the function scope, not contained in blocks), however this breaks real ES6 code where the functions might refer to block-scoped lets, consts and classes, where declarations within blocks are (finally) permitted.

I will try and synthesize a method of working out which approach to take - probably by checking to see if the function body being analyzed uses any block-scoped values.

**Comment by epoberezkin** [comment]
I think the safest and probably the easiest is to put them in the end of the function (in the same way you do vars) so just remember them when they are created and add to the end. As long as the names are unique (and they seem to be unique anyway) that should work in all cases (ES5/6 strict/non-strict).

**Comment by matAtWork** [comment]
The issue is the following case:
````
async function x(f) {
    if (f == 1) {
        let j = 1;
        if (f) {
            await afn(j);
        }
        return j;
    }
}

````
which compiles to:
````
function x(f) {
    return new Promise((function ($return, $error) {
        if (f == 1) {
            let j;
            j = 1;
            function $IfStatement_2() {
                return $return(j);
            }
            
            if (f) {
                return afn(j).then((function ($await_3) {
                    return $IfStatement_2.call(this);
                }).$asyncbind(this, $error), $error);
            }
            return $IfStatement_2.call(this);
        }
        return $return();
    }).$asyncbind(this));
}

````
If I hoist (or push to the bottom of the function) the 'continuation' IfStatement_2, it refers to `j` which is now out of scope. In ES6, it should be either left in place (so the JS engine can hoist it to Block scope) or hoisted to the top of the block. This is permitted in ES5 (sloppy), but a syntax error in ES5-strict. If I can't think of a way to work it out, one would be to make an assumption based on the code (does it use `let` `const` or `class`) and issue a warning

**Comment by epoberezkin** [comment]
`let` means it's ES6, so can be left in place... Otherwise it can be pushed to the bottom of `function ($return, $error) {...`. Interesting, how regenerator handles it?

**Comment by matAtWork** [comment]
That's my best guess so far (check for ES6 block types and leave in place, or assume ES5 and push to the bottom). It's complicated as Chrome/V8 warns about needing 'use strict' for scoped declarations to work, although I can't find a definitive description of what V8 actual does in each case.

**Comment by matAtWork** [comment]
Incidentally, this explains why your original examples behave differently in Node v0.10/12 and 4/5. The earlier ones implement ES5-strict semantics (no block-scoped function declarations) whereas the latter implement ES6 semantics (function declarations are block-scoped, like let, const & class). It's not whether it's hoisted or pushed to the bottom, it's because in your first example the declaration is in a different block to the reference.

**Comment by matAtWork** [comment]
To complicate it more, it looks like sloppy ES6 is coming to Chrome too: https://www.chromestatus.com/feature/5891733057437696

**Comment by matAtWork** [comment]
Fixed in 2.3.11

**Comment by STRML** [comment]
I'm seeing this pop up again, but only iOS 8 seems to have an issue with it.

[Example](http://nodent.mailed.me.uk/#%2F%2F%20%40flow%0A%0A%2F%2F%20Hack%20to%20avoid%20dynamic%20require%20which%20creates%20a%20ton%20of%20tiny%20webpack%20bundles%0Aasync%20function%20getLocaleData(locale)%20%7B%0A%20%20const%20baseLocale%20%3D%20getBaseLocale(locale)%3B%0A%20%20let%20localeData%3B%0A%20%20if%20(baseLocale%20%3D%3D%3D%20'zh')%20%7B%0A%20%20%20%20localeData%20%3D%20await%20System.import(%60react-intl%2Flocale-data%2Fzh.js%60)%3B%0A%20%20%7D%20else%20if%20(baseLocale%20%3D%3D%3D%20'en')%20%7B%0A%20%20%20%20localeData%20%3D%20await%20System.import(%60react-intl%2Flocale-data%2Fen.js%60)%3B%0A%20%20%7D%20else%20%7B%0A%20%20%20%20throw%20new%20Error(%22Unknown%20locale%20for%20react-intl%20load%3A%20%22%20%2B%20baseLocale)%3B%0A%20%20%7D%0A%20%20%2F%2F%20This%20loads%20locale%20data%20into%20react-intl.%0A%20%20addLocaleData(localeData)%3B%0A%7D%0A~options~%7B%22mode%22%3A%22lazyThenables%22%2C%22promiseType%22%3A%22Zousan%22%2C%22noRuntime%22%3Afalse%2C%22es6target%22%3Afalse%2C%22wrapAwait%22%3Afalse%2C%22spec%22%3Afalse%7D)

The offending bits:

```js
if (baseLocale === 'zh') {
  // ..
} else {
  //..
  function $If_2() {
      return $If_1.call(this);
  }
  
  return $If_2.call(this);
}
```

**Comment by matAtWork** [comment]
This is very specific case (and complicated) case. The reason is, in ES6, function declarations are block-scoped. In ES5, they are function scoped. Most ES5 parser allowed you to put function declarations in a nested block and simply hoisted them to function scoped (like `var` does), however, in `strict` mode some (notably Safari) throw this out as an error, since strictly speaking it was never syntactically legal to declare a function in a nested block. There is a similar ambiguity with `const` (which is not strictly defined in ES5, but is a de facto standard).

In order to generate syntactically correct code that follows these rules, a transpiler needs to know whether the code is destined for ES6 or ES5 and whether it is `strict`, since functions in blocks are illegal ES5-strict mode.

Nodent can tell if a function is `strict`, but it can't tell if it's destined for ES5 or ES6, so in an attempt to do this, it checks if you code is using ES6-like syntax. In you example, you're using string-templates, which are an ES6 feature, so it does NOT hoist block-scoped function declarations as these are legal (and meaningful in ES6). In fact, hoisting these functions in ES6 can break stuff as `let` and `const` are also block-scoped.

Take a look here: http://nodent.mailed.me.uk/#async%20function%20strict(x)%20%7B%0A%20%20%20%20%22use%20strict%22%3B%0A%20%20%20%20var%20y%3B%0A%20%20%20%20if%20(x%20%3C%200)%20%7B%0A%20%20%20%20%20%20%20%20y%20%3D%20await%201%3B%0A%20%20%20%20%7D%20else%20if%20(x%20%3E%200)%20%7B%0A%20%20%20%20%20%20%20%20y%20%3D%20await%202%3B%0A%20%20%20%20%7D%0A%7D%0A%0A

This is your code path with only ES5 syntax in `strict` mode. Nodent assumes (from the code) that there's no such thing as a block-scoped function declaration (which there isn't in ES5) and hoisted the `function $If_2()` to the enclosing function scope, which is strict, correct ES5, and will run in Safari.

Now change the `var y` to `let y`. The declaration of `function $If_2()` moves back to its original position since in ES6 functions are block-scoped, and can access block-scoped variables - hoisting it would break code.

Note that the hoisting of this kind of block-scoped function is only logically necessary in ES5, and only carried out by Nodent if you specify `strict` (go back to `var` and remove the `use strict` to see it), since all ES5 parsers I could find (maybe not Safari on iOS 8?) didn't raise an error unless `strict` is used.

So, to summarize, the issues is that:
* You're using ES6-like constructs and running in an ES5 environment.
* Nodent does NOT apply ES5-strict function hoisting (because your code contains ES6 syntax)

There's a perfectly valid (and different) issue about optimizing away the function call which just calls one function, which I might have a go at some time.

**Comment by MatAtBread** [comment]
...and I suppose the other way you could resolve this is to use Babel _first_ to generate correct ES5 code, then use nodent (or fast-async) as the last step. In this case, nodent won't think you're in an ES6 environment (since they'll be no ES6 syntax) and will hoist correctly for ES5 strict mode

**Comment by STRML** [comment]
Thanks for the comprehensive responses. The issue is indeed with `fast-async` assuming that I'm in an ES6 environment, which I'm not. It's tricky, though, because block-scoped vars [are emulated](https://babeljs.io/repl/#?babili=false&evaluate=true&lineWrap=true&presets=es2015%2Cstage-2&code=if%20(true)%20%7B%0A%20%20let%20foo%20%3D%20'bar'%3B%0A%7D%0Aconsole.log(foo)&experimental=true&loose=true&spec=false&playground=false) in babel so this could break things.

It seems the only way to merge the two concepts is to use function expressions rather than declarations. It's a deopt to be sure, so perhaps it should be opt-in, but IMO in `spec` mode it is correct to leave it on.

**Comment by MatAtBread** [comment]
It would seem that one solution would be for Babel to enforce block scoping of function declarations, since this is an ES6 feature that doesn't exist in ES5. I seem to recall there is a Babel plugin that does this. Running that after fast-async should resolve the issue. If I can find the name of the plugin I'll let you know

**Comment by STRML** [comment]
https://github.com/babel/babel/tree/master/packages/babel-plugin-transform-es2015-block-scoped-functions perhaps?

**Comment by matAtWork** [comment]
FYI, I had a quick play on the Babel playground, and if you pass the output of the nodent fragment above (either with or without the `let`), it generates correct ES5 code, so the issue is probably solved by making sure fast-async is configured to be the first plugin, before all other ES2015 transformers. In the case of block-scoped strict mode functions, it transforms them into legal function expressions, which can  (of course) be situated in blocks.

**Comment by STRML** [comment]
Just realized this is a dupe of https://github.com/MatAtBread/fast-async/issues/30 - thanks for the help.

**Comment by van-nguyen** [comment]
I'm using rollup and buble and have major issues with this one. Has anyone found a solution for this specific case?
